pkg/utils/customresource/typegenerics.go

```diff
@@ -0,0 +1,49 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package customresource
+
+import (
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+)
+
+// RetrieveLatestUpdateTime takes an arbitrary custom resource and retrives its latest update time
+// based on the resource's metadata
+func RetrieveLatestUpdateTime(customResource interface{}) (*time.Time, error) {
+	metaAccessor, maAssertable := customResource.(metav1.ObjectMetaAccessor)
+	resourceKR, krAssertable := customResource.(duckv1.KRShaped)
+	if !maAssertable || !krAssertable {
+		return nil, fmt.Errorf("The specified customResource does not support an expected type assertion (ObjectMetaAccessor: %v, KRShaped: %v) %#v", maAssertable, maAssertable, customResource)
+	}
+	latestUpdateTime := metaAccessor.GetObjectMeta().GetCreationTimestamp()
+	if metaAccessor.GetObjectMeta().GetDeletionTimestamp() != nil {
+		latestUpdateTime = *metaAccessor.GetObjectMeta().GetDeletionTimestamp()
+	} else {
+		status := resourceKR.GetStatus()
+		if status != nil && status.Conditions != nil {
+			for _, cond := range status.Conditions {
+				if latestUpdateTime.Before(&cond.LastTransitionTime.Inner) {
+					latestUpdateTime = cond.LastTransitionTime.Inner
+				}
+			}
+		}
+	}
```

Comment:
What's your thought on how we should read the metric? I can think of two confusing case.

1. A trigger's latestUpdateTime was updated before we report the metric here. It would give us an illusion of shorter delays.
2. There is no update to a trigger for a long time. Would the brokercell controller still gets informed on resync period? If so, on every resync we would get longer and longer delays because latestUpdateTime of the trigger is not changing.

Reply:
I think the resync only re-enqueues the brokercell, not the trigger. The trigger controller also has a resync. But I think it will only invoke the brokercell if there is actually an update (we should expect most resync to be noop).

Reply:
Thanks for clarifying on re-queuing. In general, I do expect that there will be some noise in the data. This could have been reduced if we had a clear last-update timestamp on the object. But that’s not the case, thus here is the workaround. 
Considering our goals, I think it’s ok that some of the samples won’t be precise — like in other statistical experiments, with large enough sampling and a measure more robust to outliers (e.g. percentiles/median vs mean), I expect the numbers to become more representative of the actual delays, which will help us confirm/rule out this as a root cause.